crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -585,22 +585,31 @@ def cancel
         upgrade_status.initialize_state
       end
 
+      def upgrade_pacemaker_cluster(cluster_env)
+        founder = ::Node.find(
+          "pacemaker_founder:true AND " \
+          "pacemaker_config_environment:#{cluster_env}"
+        ).first
+        upgrade_cluster founder, cluster_env
+      end
+
       # Take the list of elements as argument (that could be both nodes and clusters)
       def upgrade_nodes_disruptive(elements_to_upgrade)
         elements_to_upgrade.each do |element|
           # If role has a cluster assigned ugprade that cluster (reuse the non-disruptive code here)
           if ServiceObject.is_cluster?(element)
             cluster = ServiceObject.cluster_name element
             cluster_env = "pacemaker-config-#{cluster}"
-            founder = ::Node.find(
-              "pacemaker_founder:true AND " \
-              "pacemaker_config_environment:#{cluster_env}"
-            ).first
-            upgrade_cluster founder, cluster_env
+            upgrade_pacemaker_cluster cluster_env
           else
-            # if role has single node(s) assigned upgrade those nodes (optionally all at once)
-            # FIXME: could we do it in paralel?
-            upgrade_one_node element
+            node = ::Node.find_by_name(element)
+            if node["run_list_map"].key? "pacemaker-cluster-member"
```

Comment:
IMO this code would better fit into ```upgradable_elements_of_proposals```. Where you would "just" replace the nodenane with the cluster element before inserting it into the result list.

And if you already load the nodes in ```upgradable_elements_of_proposals``` anyway you could also rework the check for computes a bit to also just look at the run_list_map.